[Data] Disallow min_rows_per_file with partitioned parquet writes (#63368)
ray-project/ray: doc/source/data/saving-data.rst

Writing into Partitioned Dataset

> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer: **Heading case.** The style guide asks for [sentence case in every heading](https://docs.ray.io/en/master/ray-contribute/writing-style.html#headings), and task headings in the imperative. This one predates your PR, so it's genuinely optional — but it's inside your hunk and nothing in the tree cross-references it (no `:ref:` targets it and `autosectionlabel` isn't enabled in `doc/source/conf.py`), so it's a safe rename. The replacement is the same character count, which keeps the RST underline valid.

```suggestion
Write into a partitioned dataset
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

When writing partitioned dataset (using Hive-style, folder-based partitioning), it's
recommended to repartition the dataset by the partition columns prior to writing into it.
This lets you control the file sizes and their number. When the dataset is repartitioned
by the partition columns, every block should contain all of the rows corresponding to a
particular partition, meaning that the number of files created should be controlled by
the repartitioning, with optional limits from the write method, such as
``max_rows_per_file``. Since every block is written out independently, when writing the
dataset without prior repartitioning you could potentially get an N number of files per
partition (where N is the number of blocks in your dataset) with very limited ability to
control the number of files and their sizes, since every block could potentially carry

> the rows corresponding to any partition.

Reviewer: **Paragraph rewrite.** A few style-guide items land on the lines you rewrote here:

- `lets` is on the [avoid list](https://docs.ray.io/en/master/ray-contribute/writing-style.html#focus-on-what-the-reader-does) alongside `allows` and `enables` — the original said "allows you to," so this one swapped one flagged verb for another.
- `prior to` → `before` ([prefer plain words](https://docs.ray.io/en/master/ray-contribute/writing-style.html#prefer-plain-words)).
- "it's recommended to" hedges guidance that should be a [direct instruction](https://docs.ray.io/en/master/ray-contribute/writing-style.html#turn-hedging-into-direct-advice).
- Passive voice on "is repartitioned," "be controlled," and "is written" — these are the three findings Vale reports on this file.
- Missing article in "When writing partitioned dataset."
- "an N number of files" → "N files."
- The guide asks you to [avoid parentheticals](https://docs.ray.io/en/master/ray-contribute/writing-style.html#sentence-structure) and to split long sentences rather than joining clauses.

No technical content changes:

```suggestion
When you write a partitioned dataset using Hive-style, folder-based partitioning,
repartition the dataset by the partition columns first. Repartitioning gives you control
over the number of files and their sizes. After you repartition by the partition columns,
every block holds all the rows for a particular partition, so the repartitioning
determines how many files Ray creates, with optional limits from the write method such as
``max_rows_per_file``. Ray writes every block out independently, so if you write the
dataset without repartitioning first, you can get N files per partition, where N is the
number of blocks in your dataset. In that case you have very limited control over the
number of files and their sizes, because every block can carry rows for any partition.
```

**Optional, and outside what I'd push for:** the style guide asks contributors to [soft-wrap prose](https://docs.ray.io/en/master/ray-contribute/writing-style.html#soft-wrap-prose) — one source line per paragraph — so diffs stay clean. That rule sits under "MyST and Markdown specifics" and the rest of this `.rst` file is hard-wrapped, so I kept your wrap width above rather than making this paragraph the odd one out. Entirely your call.

.. warning::
    Ray Data deprecates using ``min_rows_per_file`` with non-empty ``partition_cols``
    and plans to stop supporting this combination after February 2027. Use
    ``repartition()`` with the partition columns and an explicit ``num_blocks``, and use
    ``max_rows_per_file`` instead. If the dataset is already repartitioned by the

>     partition columns, removing ``min_rows_per_file`` doesn't change the output layout.

Reviewer: **Admonition wording.** Two style-guide points, and one readability nit:

- The contractions rule carves out an explicit exception: ["Avoid contractions in a warning or an error message, where the extra weight of the full form helps."](https://docs.ray.io/en/master/ray-contribute/writing-style.html#use-contractions) So `doesn't` → `does not` here specifically.
- "Ray Data deprecates using X ... and plans to stop supporting" reads as intent rather than behavior. "Support ... ends after February 2027" states the same fact without the hedge, and keeps [present tense](https://docs.ray.io/en/master/ray-contribute/writing-style.html#write-in-present-tense).
- Splitting the first sentence in two makes the deprecation and the end date each scannable on their own.

**The February 2027 date, the `num_blocks` guidance, and the "output layout doesn't change" claim are unchanged** — those are @AarryaSaraf's, and I'm not restating them more strongly than the current text does.

```suggestion
.. warning::
    Using ``min_rows_per_file`` with non-empty ``partition_cols`` is deprecated. Support
    for this combination ends after February 2027. Instead, call ``repartition()`` with
    the partition columns and an explicit ``num_blocks``, and use ``max_rows_per_file``.
    If you already repartition the dataset by the partition columns, removing
    ``min_rows_per_file`` does not change the output layout.
```

>     #    block – this minimises shuffling during the write.

Reviewer: **Code comment, two small things.** `minimises` is British spelling; Ray docs follow US spelling through the Google style guide. The `–` is a Unicode en dash, and the guide asks you to [avoid dashes in prose and restructure instead](https://docs.ray.io/en/master/ray-contribute/writing-style.html#sentence-structure).

This line is context rather than something you changed, so skip it freely if you'd rather keep the diff tight.

```suggestion
    #    block. This minimizes shuffling during the write.
```

**Two more in this snippet, both outside the diff hunks, so there's no suggestion to click:**

- Line 199, `# Sample dataset that we'll partition by ``city`` and ``year``.` — the guide [rules out first-person plural](https://docs.ray.io/en/master/ray-contribute/writing-style.html#avoid-first-person-plural) even in code comments, and that apostrophe is a Unicode right single quote. "Sample dataset to partition by ``city`` and ``year``." fixes both.
- Line 216 ends `city=<value>/year=<value>/....` with four dots. Three is the ellipsis.

Both predate this PR. Mentioning them in case you're in the file anyway.
